python/ray/tune/execution/tune_controller.py

```diff
@@ -338,43 +336,39 @@ def checkpoint_exists(cls, directory: str) -> bool:
         return _experiment_checkpoint_exists(directory)
 
     def save_to_dir(self):
-        """Save TuneController state to the local experiment directory.
+        """Save TuneController state to the local staging experiment directory.
 
         This includes:
         - trial states
         - TuneController internal state (all the serializable attributes)
         - the searcher state
         - the callback states
         """
-        experiment_dir = self._storage.experiment_local_path
-
         # Get state from trial executor and runner
         runner_state = {
             # Trials
             "trial_data": list(self._get_trial_checkpoints().values()),
             # Experiment data
             "runner_data": self.__getstate__(),
             # Metadata
-            "stats": {
-                "start_time": self._start_time,
-                "timestamp": self._last_checkpoint_time,
-            },
+            "stats": {"start_time": self._start_time},
         }
-
-        tmp_file_name = Path(
-            experiment_dir, f".tmp_experiment_state_{uuid.uuid4()}"
+        experiment_state_fs_path = Path(
+            self._storage.experiment_fs_path, self.experiment_state_file_name
         ).as_posix()
 
-        with open(tmp_file_name, "w") as f:
-            json.dump(runner_state, f, indent=2, cls=TuneFunctionEncoder)
+        with self._storage.storage_filesystem.open_output_stream(
+            experiment_state_fs_path
+        ) as f:
+            f.write(json.dumps(runner_state, cls=TuneFunctionEncoder).encode("utf-8"))
 
-        os.replace(
-            tmp_file_name,
-            Path(experiment_dir, self.experiment_state_file_name).as_posix(),
+        experiment_local_staging_path = self._storage.experiment_local_staging_path
+        self._search_alg.save_to_dir(
+            experiment_local_staging_path, session_str=self._session_str
+        )
+        self._callbacks.save_to_dir(
+            experiment_local_staging_path, session_str=self._session_str
         )
```

Comment:
This change uploads just the TuneController to state to storage directly, rather than dumping it to a file and waiting for the driver sync to upload it.

Since searcher state saving is really complicated, I still dump those files and upload them implicitly through driver sync uploading.

The reason I made this change is to fix some broken tests that weren't uploading the `TuneController` state often enough. The local "staging" version of the file was up to date, but it wasn't being uploaded in time to restore from the right state after a crash on the driver (ex: a callback raised an error).

Reply:
This solution has been replaced with dumping to the directory + uploading immediately after.

This is because searcher state is too hard to refactor -- the implementation of save for many searchers is tied to writing files.

See https://github.com/ray-project/ray/pull/43403/files#diff-56de03a86c3b7df66c61730aba6813a52bf26cbb424357b9bd35b7294b18294bR185-R189